serverless/serverless
# [WIP] AWS: Attach Lambdas to existing Cognito User Pool

## What did you implement:

Partially addresses #4241.
Closes #4207.
Closes #5401 

## How did you implement it:

Add `preExisting` field to `cognitoUserPool` event.

If `preExisting` is not given or `false`, then `pool` field is a pool **name*, as same as before this PR.

If `preExisting` is `true`, then `pool` is expected to be a **pool Id** or an **ARN** of an existing user pool
In this case, new Cognito User Pool resource is not created and just referenced.
Lambdas are attached/removed as triggers to CUP via AWS SDK on hooks, not via CloudFormation, since CloudFormation does not support configure pre-existing CUP.

## How can we verify it:

1. `npm install -g exoego/serverless#aws-existing-resources`
2. Prepare serverless.yaml based onf the below
```
service: pre-existing

provider:
  name: aws
  runtime: nodejs8.10
  region: us-east-1

functions:
  hello:
    handler: handler.hello
    events: []
      - cognitoUserPool:
          pool: arn:aws:cognito-idp:REGION_HERE:ACCOUNT_ID_HERE:userpool/POOL_ID_HERE
          trigger: PreSignUp
          preExisting: true
```
3. `sls deploy`
4. Go to the Trigger page of the specified user pool on Cognito Console, then confirm that `PreSignUp` trigger is now `pre-existing-dev-hello` lambda.
5. `sls remove`
6. Refresh the Trigger page, then confirm that `PreSignUp` trigger is now empty.

## Todos:


***Is this ready for review?:*** NO
***Is it a breaking change?:*** NO
